File: sdk/nestconnect_flutter/lib/src/voice.dart
import 'dart:async';
import 'dart:math' as math;

import 'package:flutter/material.dart';
import 'package:just_audio/just_audio.dart';
import 'package:nestconnect_client/nestconnect_client.dart';

import 'theme.dart';

/// The bars a note carries, and what to draw when it carries none.
///
/// Deliberately uneven. A row of identical bars reads as a loading state
/// rather than as audio.
const _fallbackBars = <double>[0.3, 0.6, 0.4, 0.8, 0.5, 0.7, 0.35, 0.65, 0.45, 0.55];

/// "0:07" — what a bubble says under a note.
String formatDuration(Duration d) {
  final total = d.inSeconds < 0 ? 0 : d.inSeconds;
  final seconds = (total % 60).toString().padLeft(2, '0');
  return '${total ~/ 60}:$seconds';
}

/// Squash however many samples were taken into the bars a bubble draws.
///
/// Peak per bucket, not mean, and that is the whole difference between a
/// waveform that looks like speech and one that looks like a hedge. Speech is
/// mostly quiet — the gaps between words are real silence — so averaging pulls
/// every bar toward the middle and produces an even block. The loudest moment
/// in each bucket keeps the syllables.
List<double> toBars(List<double> samples, {int bars = 60}) {
  if (samples.isEmpty) return const [];
  if (samples.length <= bars) {
    return samples.map((v) => v.clamp(0.0, 1.0)).toList(growable: false);
  }
  final out = <double>[];
  final per = samples.length / bars;
  for (var i = 0; i < bars; i++) {
    final from = (i * per).floor();
    final to = math.max(from + 1, ((i + 1) * per).floor());
    var peak = 0.0;
    for (var j = from; j < to && j < samples.length; j++) {
      peak = math.max(peak, samples[j]);
    }
    out.add(peak.clamp(0.0, 1.0));
  }
  return out;
}

/// Lift the quiet parts so a normal speaking voice fills the bubble.
///
/// A phone microphone at arm's length reads low, and drawn honestly that is a
/// flat line with two bumps. Every messenger normalises this: the note is not
/// a measurement, it is a picture of somebody talking, and it should look like
/// they were there. Floored, so near-silence draws as a thin line rather than
/// as nothing at all.
List<double> normalise(List<double> samples) {
  if (samples.isEmpty) return const [];
  final peak = samples.fold<double>(0, math.max);
  if (peak <= 0) return List<double>.filled(samples.length, 0.06);
  return samples
      .map((v) => math.max(0.06, math.pow(v / peak, 0.7).toDouble()).clamp(0.0, 1.0))
      .toList(growable: false);
}

/// Turn the recorder's amplitude, in dBFS, into 0..1.
///
/// `record` reports decibels relative to full scale: 0 is as loud as the
/// hardware goes and anything quiet is a large negative number. −45 dB is
/// about the noise floor of a room, so that is treated as silence rather than
/// mapping the whole range and drawing a bubble full of hiss.
double levelFromDb(double db) {
  if (!db.isFinite) return 0;
  const floor = -45.0;
  if (db <= floor) return 0;
  return ((db - floor) / -floor).clamp(0.0, 1.0);
}

/// The bars themselves — shared by the player and the live meter, so a note
/// looks the same while it is being recorded and after it has been sent.
class NestWaveform extends StatelessWidget {
  const NestWaveform({
    super.key,
    required this.bars,
    required this.color,
    this.playedColor,
    this.progress = 0,
    this.height = 26,
  });

  final List<double> bars;
  final Color color;

  /// The part already played. Null while recording, where nothing is behind.
  final Color? playedColor;
  final double progress;
  final double height;

  @override
  Widget build(BuildContext context) {
    final shown = bars.isEmpty ? _fallbackBars : bars;
    return SizedBox(
      height: height,
      child: Row(
        crossAxisAlignment: CrossAxisAlignment.center,
        children: [
          for (var i = 0; i < shown.length; i++)
            Expanded(
              child: Padding(
                padding: const EdgeInsets.symmetric(horizontal: 1),
                child: AnimatedContainer(
                  // Short enough to track a voice, long enough that the bars
                  // glide between samples rather than snapping.
                  duration: const Duration(milliseconds: 90),
                  height: math.max(3, shown[i] * height),
                  decoration: BoxDecoration(
                    // Each bar colours itself by comparing its own position
                    // with the playhead. An overlay would have to clip
                    // mid-bar, and would shimmer as it crossed one.
                    color: playedColor != null && i / shown.length < progress
                        ? playedColor
                        : color,
                    borderRadius: BorderRadius.circular(2),
                  ),
                ),
              ),
            ),
        ],
      ),
    );
  }
}

/// A voice note, played in the bubble it arrived in.
///
/// Nothing is fetched until play is pressed: a thread with ten notes in it
/// would otherwise pull ten files on open, most of which nobody listens to,
/// over a connection that is usually a phone's.
class NestVoiceNote extends StatefulWidget {
  const NestVoiceNote({
    super.key,
    required this.attachment,
    required this.source,
    required this.mine,
    required this.theme,
  });

  final NestAttachment attachment;

  /// Where to play it from — a URL on the server, or a file still on its way.
  final Uri? source;
  final bool mine;
  final NestTheme theme;

  @override
  State<NestVoiceNote> createState() => _NestVoiceNoteState();
}

class _NestVoiceNoteState extends State<NestVoiceNote> {
  AudioPlayer? _player;
  StreamSubscription<Duration>? _ticks;
  StreamSubscription<PlayerState>? _states;
  Duration _at = Duration.zero;
  bool _playing = false;
  bool _failed = false;

  Duration get _total => Duration(milliseconds: widget.attachment.durationMs ?? 0);

  @override
  void dispose() {
    unawaited(_ticks?.cancel());
    unawaited(_states?.cancel());
    unawaited(_player?.dispose());
    super.dispose();
  }

  Future<void> _toggle() async {
    final src = widget.source;
    if (src == null || _failed) return;

    if (_playing) {
      await _player?.pause();
      if (mounted) setState(() => _playing = false);
      return;
    }

    try {
      final player = _player ??= AudioPlayer();
      if (_player != null && _ticks == null) {
        _ticks = player.positionStream.listen((p) {
          if (mounted) setState(() => _at = p);
        });
        _states = player.playerStateStream.listen((s) {
          if (!mounted) return;
          if (s.processingState == ProcessingState.completed) {
            // Back to the start, so the next press replays. A player left
            // sitting at its own end is the commonest way a second tap
            // appears to be ignored.
            setState(() {
              _playing = false;
              _at = Duration.zero;
            });
            unawaited(player.seek(Duration.zero));
            unawaited(player.pause());
          } else {
            setState(() => _playing = s.playing);
          }
        });
      }
      if (player.audioSource == null) await player.setUrl(src.toString());
      await player.play();
    } catch (_) {
      // A note that will not load is said so rather than left as a button
      // that does nothing when pressed.
      if (mounted) setState(() => _failed = true);
    }
  }

  Future<void> _seek(double ratio) async {
    final ms = (_total.inMilliseconds * ratio).round();
    setState(() => _at = Duration(milliseconds: ms));
    await _player?.seek(Duration(milliseconds: ms));
  }

  @override
  Widget build(BuildContext context) {
    final t = widget.theme;
    final foreground = widget.mine ? t.onMine : t.onTheirs;
    final progress = _total.inMilliseconds == 0
        ? 0.0
        : (_at.inMilliseconds / _total.inMilliseconds).clamp(0.0, 1.0);
    // While playing, the player's own clock: a note whose stored duration
    // rounded should still have its line reach the end as the sound stops.
    final shown = _playing || _at > Duration.zero ? _at : _total;

    return Container(
      constraints: BoxConstraints(maxWidth: MediaQuery.of(context).size.width * 0.72),
      padding: const EdgeInsets.fromLTRB(8, 8, 14, 8),
      decoration: BoxDecoration(
        color: widget.mine ? t.mine : t.theirs,
        borderRadius: BorderRadius.only(
          topLeft: const Radius.circular(18),
          topRight: const Radius.circular(18),
          bottomLeft: Radius.circular(widget.mine ? 18 : 6),
          bottomRight: Radius.circular(widget.mine ? 6 : 18),
        ),
      ),
      child: Row(
        mainAxisSize: MainAxisSize.min,
        children: [
          Semantics(
            button: true,
            label: _playing ? 'Pause voice message' : 'Play voice message',
            child: InkWell(
              onTap: _toggle,
              customBorder: const CircleBorder(),
              child: Container(
                width: 34,
                height: 34,
                decoration: BoxDecoration(
                  color: foreground.withValues(alpha: 0.18),
                  shape: BoxShape.circle,
                ),
                child: Icon(
                  _playing ? Icons.pause_rounded : Icons.play_arrow_rounded,
                  size: 20,
                  color: foreground,
                ),
              ),
            ),
          ),
          const SizedBox(width: 9),
          Flexible(
            child: LayoutBuilder(
              builder: (context, box) => GestureDetector(
                // The whole strip is the timeline. Scrubbing by dragging is
                // what a thumb reaches for, and a note is too short for a
                // separate slider to be worth the room.
                onTapDown: (d) => unawaited(_seek(d.localPosition.dx / box.maxWidth)),
                onHorizontalDragUpdate: (d) =>
                    unawaited(_seek((d.localPosition.dx / box.maxWidth).clamp(0.0, 1.0))),
                behavior: HitTestBehavior.opaque,
                child: NestWaveform(
                  bars: widget.attachment.waveform,
                  color: foreground.withValues(alpha: 0.35),
                  playedColor: foreground,
                  progress: progress,
                ),
              ),
            ),
          ),
          const SizedBox(width: 9),
          Text(
            _failed ? 'Unavailable' : formatDuration(shown),
            style: TextStyle(
              fontSize: 11.5,
              color: foreground.withValues(alpha: 0.75),
              fontFeatures: const [FontFeature.tabularFigures()],
            ),
          ),
        ],
      ),
    );
  }
}

File: apps/web/src/widget/voice.ts
import { NESTCHAT_VOICE_MAX_MS, NESTCHAT_WAVEFORM_BARS } from "@ding/schemas";

/**
 * Recording a voice note in a browser.
 *
 * Two things happen at once and only one of them is obvious. The obvious one
 * is MediaRecorder collecting compressed audio. The other is that the bars
 * have to be drawn *while the person is talking* — a waveform that appears
 * only after they let go reads as a progress bar, and the whole reason the
 * shape is there is to say "it is hearing you" during the part where they
 * cannot tell.
 *
 * So the same stream is also fed to an AnalyserNode, sampled on an interval,
 * and the peaks kept. What is drawn live and what is sent afterwards are the
 * same numbers, which is why the note you sent looks like the one you watched
 * yourself record.
 */

/** How often the level is sampled while recording. Fast enough that a syllable
 *  moves the bars, slow enough not to sit on the main thread. */
const SAMPLE_MS = 60;

/** What the browser is willing to record, best first. Safari cannot do webm at
 *  all and mp4 is the one it has; Chrome and Firefox prefer opus, which is
 *  dramatically smaller for speech. */
const FORMATS = [
  "audio/webm;codecs=opus",
  "audio/webm",
  "audio/mp4",
  "audio/mpeg",
] as const;

export interface VoiceNote {
  blob: Blob;
  durationMs: number;
  /** 0..1, at most {@link NESTCHAT_WAVEFORM_BARS} of them. */
  waveform: number[];
}

/** The type this browser will actually record, or undefined if none. */
export function pickMimeType(
  supported: (type: string) => boolean = (t) =>
    typeof MediaRecorder !== "undefined" && MediaRecorder.isTypeSupported(t),
): string | undefined {
  return FORMATS.find((f) => {
    try {
      return supported(f);
    } catch {
      return false;
    }
  });
}

/**
 * Squash however many samples were taken into the bars a bubble draws.
 *
 * Peak rather than mean, and this is the whole difference between a waveform
 * that looks like speech and one that looks like a hedge. Speech is mostly
 * quiet — the gaps between words are real silence — so averaging a bucket
 * pulls every bar toward the middle and produces an even green block. Taking
 * the loudest moment in each bucket keeps the syllables.
 *
 * A short note is padded with the samples it has rather than stretched: six
 * bars for a one-second note is honest, and sixty bars interpolated from six
 * is a drawing of nothing.
 */
export function toBars(samples: number[], bars = NESTCHAT_WAVEFORM_BARS): number[] {
  if (!samples.length) return [];
  if (samples.length <= bars) return samples.map(clamp01);
  const out: number[] = [];
  const per = samples.length / bars;
  for (let i = 0; i < bars; i++) {
    const from = Math.floor(i * per);
    const to = Math.max(from + 1, Math.floor((i + 1) * per));
    let peak = 0;
    for (let j = from; j < to && j < samples.length; j++) peak = Math.max(peak, samples[j]!);
    out.push(clamp01(peak));
  }
  return out;
}

/**
 * Lift the quiet parts so a normal speaking voice fills the bubble.
 *
 * Raw RMS off a phone microphone at arm's length sits around 0.05–0.15, which
 * drawn honestly is a flat line with a couple of bumps. Every messenger
 * normalises this; the note is not a measurement, it is a picture of somebody
 * talking, and it should look like they were there.
 *
 * Normalised against the loudest moment rather than a fixed gain, so a shout
 * and a mutter both fill the space — and floored, so a note recorded in near
 * silence draws as a thin line rather than as nothing at all.
 */
export function normalise(samples: number[]): number[] {
  const peak = samples.reduce((max, v) => Math.max(max, v), 0);
  if (peak <= 0) return samples.map(() => 0.06);
  return samples.map((v) => clamp01(Math.max(0.06, (v / peak) ** 0.7)));
}

const clamp01 = (v: number) => (Number.isFinite(v) ? Math.min(1, Math.max(0, v)) : 0);

/** How loud this moment is, 0..1, from a frame of time-domain samples. */
export function levelOf(frame: Uint8Array): number {
  if (!frame.length) return 0;
  let sum = 0;
  for (let i = 0; i < frame.length; i++) {
    // 128 is silence in a byte-valued time domain; the distance from it is
    // the amplitude either way.
    const v = (frame[i]! - 128) / 128;
    sum += v * v;
  }
  return clamp01(Math.sqrt(sum / frame.length));
}

export interface Recorder {
  /** Bars so far, already normalised — what the live meter draws. */
  readonly live: () => number[];
  readonly elapsedMs: () => number;
  /** Stop and keep it. */
  stop: () => Promise<VoiceNote | undefined>;
  /** Stop and throw it away — the slide-to-cancel. Releases the microphone
   *  exactly as `stop` does: the light going out is how somebody knows. */
  cancel: () => void;
}

/**
 * Ask for the microphone and start recording.
 *
 * Throws if permission is refused or there is no microphone, which the caller
 * turns into a message rather than a broken button. Nothing is retried: a
 * person who said no to the prompt has answered the question.
 */
export async function startRecording(opts?: {
  onTick?: (bars: number[], elapsedMs: number) => void;
  maxMs?: number;
  onMaxReached?: () => void;
}): Promise<Recorder> {
  const mimeType = pickMimeType();
  const stream = await navigator.mediaDevices.getUserMedia({
    // Three things every phone messenger turns on, and the reason a voice note
    // recorded on a bus is intelligible at all.
    audio: { echoCancellation: true, noiseSuppression: true, autoGainControl: true },
  });

  const recorder = new MediaRecorder(stream, mimeType ? { mimeType } : undefined);
  const chunks: Blob[] = [];
  recorder.ondataavailable = (e) => {
    if (e.data.size) chunks.push(e.data);
  };

  const audio = new AudioContext();
  const analyser = audio.createAnalyser();
  analyser.fftSize = 512;
  audio.createMediaStreamSource(stream).connect(analyser);
  const frame = new Uint8Array(analyser.fftSize);

  const samples: number[] = [];
  const startedAt = Date.now();
  const maxMs = opts?.maxMs ?? NESTCHAT_VOICE_MAX_MS;

  const timer = setInterval(() => {
    analyser.getByteTimeDomainData(frame);
    samples.push(levelOf(frame));
    const elapsed = Date.now() - startedAt;
    opts?.onTick?.(normalise(toBars(samples)), elapsed);
    // A cap rather than a warning. Somebody who has been holding the button
    // for five minutes has forgotten they are holding it, and the alternative
    // to stopping is a file nobody will listen to.
    if (elapsed >= maxMs) opts?.onMaxReached?.();
  }, SAMPLE_MS);

  // Released together, always — a microphone left open is the indicator light
  // staying on after the person thinks they have finished.
  const release = () => {
    clearInterval(timer);
    stream.getTracks().forEach((t) => t.stop());
    void audio.close().catch(() => undefined);
  };

  recorder.start();

  return {
    live: () => normalise(toBars(samples)),
    elapsedMs: () => Date.now() - startedAt,
    async stop() {
      const durationMs = Date.now() - startedAt;
      const done = new Promise<void>((resolve) => {
        recorder.onstop = () => resolve();
      });
      if (recorder.state !== "inactive") recorder.stop();
      await done;
      release();
      if (!chunks.length) return undefined;
      return {
        blob: new Blob(chunks, { type: mimeType ?? chunks[0]!.type }),
        durationMs,
        waveform: normalise(toBars(samples)),
      };
    },
    cancel() {
      // No onstop handler: the chunks are deliberately abandoned rather than
      // assembled. Stopping the recorder first still matters — it is what
      // flushes the encoder and lets the tracks close cleanly.
      if (recorder.state !== "inactive") {
        recorder.onstop = null;
        recorder.stop();
      }
      release();
    },
  };
}

/** "0:07". What a bubble says under a note. */
export function clockDuration(ms: number): string {
  const total = Math.max(0, Math.round(ms / 1000));
  return `${Math.floor(total / 60)}:${String(total % 60).padStart(2, "0")}`;
}
